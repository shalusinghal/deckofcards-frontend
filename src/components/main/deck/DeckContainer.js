import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Card from './card/Card';

const DeckContainer = (props) => {
  const { decks } = props;
  const renderContainer = decks.map((deck, index) => {
    const { deckId } = deck;
    return (
      <div className="col-md-6" key={deckId}>
        <h4>Deck { index + 1 }</h4>
        <div className="row">
          <Card deck={deck} />
        </div>
      </div>
    );
  });

  return (
    <div className="row mt-3">
      { renderContainer }
    </div>
  );
};

DeckContainer.propTypes = {
  decks: PropTypes.oneOfType([PropTypes.array]).isRequired,
};

const mapStateToProps = (state) => ({
  decks: state.deck.decks,
});


const mapDispatchToProps = {};

const DeckContainerConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
)(DeckContainer);

export default DeckContainerConnect;
