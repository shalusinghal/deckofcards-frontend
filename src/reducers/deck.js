import { ACTION_TYPES, CARD } from '../constants';

const prepareDecks = (action, state) => {
  const { deckIds, totalItems } = action;

  const card = {
    frontSrc: CARD.frontSrc,
    backSrc: CARD.backSrc,
    displayState: 'front',
    alreadyDrawn: false,
  };

  const newDeck = deckIds.map((deckId) => {
    let cards = [];
    // Add the card and code for each deck
    for (let index = 0; index < totalItems; index += 1) {
      cards = [...cards, { ...card, code: `${index}_${deckId}` }];
    }

    return { deckId, cards };
  });

  return {
    ...state,
    decks: newDeck,
    cardDrawCounter: 0,
    match: 0,
    lastOpenCard: null,
  };
};

const changeDisplayStateOfCard = (deck, index) => (
  deck.cards.map((card, counter) => {
    if (counter === index) {
      return { ...card, displayState: (card.displayState === 'front') ? 'back' : 'front' };
    }
    return card;
  })
);

const doToggleCard = (action, state) => {
  const { deckId, index } = action;
  const decks = [...state.decks];
  const newDecks = decks.map((deck) => {
    if (deck.deckId === deckId) {
      // Change the display state
      const cards = changeDisplayStateOfCard(deck, index);
      return { deckId, cards };
    }
    return deck;
  });

  return { ...state, decks: newDecks };
};

const setNewCard = (action, state) => {
  const { deckId, newCard, index } = action.options;
  const decks = [...state.decks];

  const newDecks = decks.map((deck) => {
    if (deck.deckId === deckId) {
      // set the new card to same index and remaiing will be same
      const newDeck = { ...deck };
      const { cards } = newDeck;
      cards[index] = newCard;

      return { deckId, cards: newDeck.cards };
    }
    return deck;
  });

  return { ...state, decks: newDecks };
};

const setLastOpenCard = (action, state) => {
  const { deckId, index } = action.options;

  // return null if last open card is already there
  if (state.lastOpenCard !== null) {
    return { ...state, lastOpenCard: null };
  }

  // Get the selected deck and card and set it as last open card
  const selectedDeck = state.decks.find((deck) => deck.deckId === deckId);
  const newLastOpenCard = selectedDeck.cards.find((card, cardIndex) => cardIndex === index);
  newLastOpenCard.deckId = deckId;
  newLastOpenCard.index = index;

  return { ...state, lastOpenCard: newLastOpenCard };
};

const initialState = {
  decks: [],
  cardDrawCounter: 0,
  match: 0,
  lastOpenCard: null,
};

const deck = (state = initialState, action) => {
  switch (action.type) {
    case ACTION_TYPES.shuffle:
      return prepareDecks(action, state);
    case ACTION_TYPES.toggleCard:
      return doToggleCard(action, state);
    case ACTION_TYPES.newDraw:
      return setNewCard(action, state);
    case ACTION_TYPES.lastOpenCard:
      return setLastOpenCard(action, state);
    case ACTION_TYPES.cardDrawCounter:
      return { ...state, cardDrawCounter: state.cardDrawCounter + 1 };
    case ACTION_TYPES.cardMatch:
      return { ...state, match: state.match + 1 };
    default:
      return state;
  }
};

export default deck;
