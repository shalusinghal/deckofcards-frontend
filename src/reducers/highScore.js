import { ACTION_TYPES } from '../constants';

const highScore = (state = { highScoreList: [] }, action) => {
  if (action.type === ACTION_TYPES.fetchHighScore) {
    return { highScoreList: action.highScoreList };
  }

  return state;
};

export default highScore;
