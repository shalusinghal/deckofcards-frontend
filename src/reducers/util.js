import { ACTION_TYPES } from '../constants';

const util = (state = { loading: false, isError: false }, action) => {
  switch (action.type) {
    case ACTION_TYPES.loading:
      return { ...state, loading: !state.loading };
    case ACTION_TYPES.error:
      return { ...state, isError: action.errorFlag };
    default:
      return state;
  }
};

export default util;
