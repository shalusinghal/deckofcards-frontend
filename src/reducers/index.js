import { combineReducers } from 'redux';

import highScore from './highScore';
import util from './util';
import deck from './deck';

const gameApp = combineReducers({
  deck,
  highScore,
  util,
});

export default gameApp;
