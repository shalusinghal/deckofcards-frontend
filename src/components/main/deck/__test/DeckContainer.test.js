import React from 'react';
import { createStore } from 'redux';
import renderer from 'react-test-renderer';
import reducers from '../../../../reducers';

import DeckContainer from '../DeckContainer';

const store = createStore(reducers);

it('renders correctly when there are no data', () => {
  const tree = renderer.create(<DeckContainer store={store} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly when there is data', () => {
  const decks = [{ deckId: '123', cards: [] }];
  const tree = renderer.create(<DeckContainer store={store} decks={decks} />).toJSON();
  expect(tree).toMatchSnapshot();
});
