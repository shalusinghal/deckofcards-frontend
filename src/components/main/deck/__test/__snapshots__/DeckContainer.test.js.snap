// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when there are no data 1`] = `
<div
  className="row mt-3"
/>
`;

exports[`renders correctly when there is data 1`] = `
<div
  className="row mt-3"
/>
`;
